### PR TITLE
[Feature] Room list filters

### PR DESCRIPTION
**Pull request by ganfra**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content
First implementation of room list filters.
The feature is behind a flag.

Note that this feature is still under development and will have iteration.
Exemple of missing bits : empty state for the room list with selected filters.

## Motivation and context
Closes https://github.com/element-hq/element-x-android/issues/2234
Closes https://github.com/element-hq/element-x-android/issues/2233
Closes https://github.com/element-hq/element-x-android/issues/2232

## Screenshots / GIFs
In PR

## Tests

- Step 1
- Step 2
- Step ...

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR

### CODE REVIEW

File: features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/RoomListView.kt [hunk 1]
```diff
@@ -204,16 +205,21 @@ private fun RoomListContent(
...
+                    onToggleSearch = { state.eventSink(RoomListEvents.ToggleSearchResults) },
+                    onMenuActionClicked = onMenuActionClicked,
+                    onOpenSettings = onOpenSettings,
+                    scrollBehavior = scrollBehavior,
+                    displayMenuItems = !state.displayMigrationStatus,
+                )
+                if (state.displayFilters) {
+                    RoomListFiltersView(state = state.filtersState)
+                }
+            }
```
Comment (bmarty) [inline features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/RoomListView.kt:222]: The filter will be displayed above the verify session banner, right? This is a bit strange, no?

Comment (ganfra) [reply features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/RoomListView.kt:222]: Yes, I think it's OK. But can review that with the design team.

File: features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersStateProvider.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            aRoomListFiltersState(
+                selectedFilters = persistentListOf(RoomListFilter.Rooms, RoomListFilter.Favourites),
+                unselectedFilters = persistentListOf(RoomListFilter.Unread),
+            ),
+        )
+}
+
+fun aRoomListFiltersState(
+    unselectedFilters: ImmutableList<RoomListFilter> = RoomListFilter.entries.toImmutableList(),
+    selectedFilters: ImmutableList<RoomListFilter> = persistentListOf(),
```
Comment (bmarty) [inline features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersStateProvider.kt:37]: Nit: for `fun aState()` methods, I found that using `List` is much simpler, and when creating the state, just call `toPersistentList()` to the parameter. Not a blocker of course!

File: features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersView.kt [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            RoomListClearFiltersButton(
+                modifier = Modifier.testTag(TestTags.homeScreenClearFilters),
+                onClick = ::onClearFiltersClicked
+            )
+        }
+        val lazyListState = rememberLazyListState()
+        val fadingEdgesBrush = horizontalFadingEdgesBrush(
+            showLeft = lazyListState.canScrollBackward,
+            showRight = lazyListState.canScrollForward
+        )
```
Comment (bmarty) [inline features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersView.kt:84]: Remark: the fading does not seem to be working on the generated preview.

File: features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersView.kt [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+@Composable
+private fun RoomListFilterView(
+    roomListFilter: RoomListFilter,
+    selected: Boolean,
+    onClick: (RoomListFilter) -> Unit,
+    modifier: Modifier = Modifier
+) {
+    FilterChip(
```
Comment (bmarty) [inline features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersView.kt:148]: Do we need to add this component in the Compound library at some point?

File: features/roomlist/impl/src/test/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersPresenterTests.kt [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    fun `present - toggle rooms filter`() = runTest {
+        val roomListService = FakeRoomListService()
+        val presenter = createRoomListFiltersPresenter(roomListService)
+        moleculeFlow(RecompositionMode.Immediate) {
+            presenter.present()
+        }.test {
+            awaitItem().eventSink.invoke(RoomListFiltersEvents.ToggleFilter(RoomListFilter.Rooms))
+
+            awaitLastSequentialItem().let { state ->
+
```
Comment (bmarty) [inline features/roomlist/impl/src/test/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersPresenterTests.kt:63]: Neat: Maybe remove those extra new lines

File: features/roomlist/impl/src/test/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersViewTests.kt [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            RoomListFiltersView(
+                state = aRoomListFiltersState(
+                    unselectedFilters = persistentListOf(),
+                    selectedFilters = RoomListFilter.entries.toImmutableList(),
+                    eventSink = eventsRecorder
+                ),
+            )
+        }
+        rule.pressTag(TestTags.homeScreenClearFilters.value)
+        eventsRecorder.assertList(
```
Comment (bmarty) [inline features/roomlist/impl/src/test/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersViewTests.kt:66]: Neat: `assertSingle` is more compact than `assertList`, maybe less extensible, but I do not think the expected result will change in the future.

File: features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/RoomListView.kt (hunk 1 above)
Comment (bmarty) [reply features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/RoomListView.kt:225]: Putting a screenshot here, for the record:

image

There is less and less space for the rooms :)

CC @amshakal

**Review by bmarty — APPROVED** [review]
Good work, thanks!
Nothing blocking, as far as this is still a disabled feature. The empty state will have to be fixed before we enable the feature, because after the login, when waiting for the room, the empty state is displayed:

image

**Comment by bmarty** [comment]
One extra question: do we want the current filter to survive an app kill and restart?

File: features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersView.kt (hunk 4 above)
Comment (ganfra) [reply features/roomlist/impl/src/main/kotlin/io/element/android/features/roomlist/impl/filters/RoomListFiltersView.kt:148]: Yes it'll probably be, but for now we can stay simple.

**Comment by ganfra** [comment]
> One extra question: do we want the current filter to survive an app kill and restart?

Not for now, but I guess it should be part of the exploration.

**Comment by ganfra** [comment]
I'll iterate on some point very soon including animations. I merge like that so we can start playing with it in real life.

**Pull request merged**